# Enforce either `Buffer` or `require("buffer").Buffer` (`n/prefer-global/buffer`)

<!-- end auto-generated rule header -->

The `Buffer` class of `buffer` module is defined as a global variable.

```js
console.log(Buffer === require("buffer").Buffer) //â†’ true
```

It will be readable if we use either `Buffer` consistently.

## ğŸ“– Rule Details

This rule enforces which `Buffer` we should use.

### Options

This rule has a string option.

```json
{
    "n/prefer-global/buffer": ["error", "always" | "never"]
}
```

- `"always"` (default) ... enforces to use the global variable `Buffer` rather than `require("buffer").Buffer`.
- `"never"` ... enforces to use `require("buffer").Buffer` rather than the global variable `Buffer`.

#### always

Examples of ğŸ‘ **incorrect** code for this rule:

```js
/*eslint n/prefer-global/buffer: [error]*/

const { Buffer } = require("buffer")
const b = Buffer.alloc(16)
```

Examples of ğŸ‘ **correct** code for this rule:

```js
/*eslint n/prefer-global/buffer: [error]*/

const b = Buffer.alloc(16)
```

#### never

Examples of ğŸ‘ **incorrect** code for the `"never"` option:

```js
/*eslint n/prefer-global/buffer: [error, never]*/

const b = Buffer.alloc(16)
```

Examples of ğŸ‘ **correct** code for the `"never"` option:

```js
/*eslint n/prefer-global/buffer: [error, never]*/

const { Buffer } = require("buffer")
const b = Buffer.alloc(16)
```

## ğŸ” Implementation

- [Rule source](../../../lib/rules/prefer-global/buffer.js)
- [Test source](../../../tests/lib/rules/prefer-global/buffer.js)
